src/Workspaces/Core/Portable/SymbolId/SymbolKey.TupleTypeSymbolKey.cs

```diff
@@ -40,19 +40,27 @@ public override SymbolKeyResolution Resolve(Compilation compilation, bool ignore
                     return SpecializedCollections.EmptyEnumerable<INamedTypeSymbol>();
                 }
 
-                if (_names == null)
+                try
                 {
-                    return SpecializedCollections.SingletonEnumerable(compilation.CreateTupleTypeSymbol(underlyingType));
+                    if (_names == null)
+                    {
+                        return SpecializedCollections.SingletonEnumerable(compilation.CreateTupleTypeSymbol(underlyingType));
+                    }
+                    else
+                    {
+                        return SpecializedCollections.SingletonEnumerable(compilation.CreateTupleTypeSymbol(underlyingType, _names.ToImmutableArray()));
+                    }
                 }
-                else
+                catch (ArgumentException)
                 {
-                    return SpecializedCollections.SingletonEnumerable(compilation.CreateTupleTypeSymbol(underlyingType, _names.ToImmutableArray()));
+                    // underlyingType is not tuple-compatible
+                    return SpecializedCollections.SingletonEnumerable(compilation.GetSpecialType(SpecialType.System_Object));
```

Comment:
I did that at first, but it ripples badly (crashes in more than one place). I don't know a way to give up the refactoring so late in the process.